apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/FakeScheduler.cs

```diff
@@ -0,0 +1,404 @@
+    /// Experimental - please read remarks. Fake implementation of <see cref="IScheduler" />, designed to work with
+    /// <see cref="FakeClock"/>.
+    /// </summary>
+    /// <remarks>
+    /// This class is currently hard to use robustly, and cancellation in particular is hard.
+    /// The API and behavior may change in future versions without a major version bump - in other words,
+    /// this is not considered part of the stable API of the package.
+    /// </remarks>
+    public sealed class FakeScheduler : IScheduler
+    {
+        /// <summary>
+        /// How long the scheduler can run in real time before timing out.
+        /// </summary>
+        public TimeSpan RealTimeTimeout { get; set; } = TimeSpan.FromSeconds(10);
+
+        /// <summary>
+        /// How long the scheduler can run in simulated time before timing out.
+        /// </summary>
+        public TimeSpan SimulatedTimeTimeout { get; set; } = TimeSpan.FromDays(10);
+
+        /// <summary>
+        /// When running the scheduler, we pause until no more tasks have been scheduled
+        /// for a certain time, to allow continuations to execute before the clock is advanced.
+        /// By default this is 10ms, but tests which expect a greater amount of processing
+        /// may increase this.
+        /// </summary>
+        public TimeSpan IdleTimeBeforeAdvancing { get; set; } = TimeSpan.FromMilliseconds(10);
+
+        /// <summary>
+        /// Time to allow test code to execute continuations after the scheduler loop has completed.
+        /// This is a one-time delay per RunAsync/RunAndPause call, to reduce flakiness.
+        /// </summary>
+        public TimeSpan PostLoopSettleTime { get; set; } = TimeSpan.FromMilliseconds(100);
+
+        /// <summary>
+        /// The clock associated with this scheduler. The clock is advanced as the scheduler runs.
+        /// </summary>
+        public FakeClock Clock { get; }
+
+        private readonly object _monitor = new object();
+        private LinkedList<DelayTimer> _actions = new LinkedList<DelayTimer>();
+        private bool _stopped;
+
+        /// <summary>
+        /// Constructs a fake scheduler which works with the given clock.
+        /// </summary>
+        /// <param name="clock">Fake clock to observe for timing purposes.</param>
+        public FakeScheduler(FakeClock clock)
+        {
+            Clock = GaxPreconditions.CheckNotNull(clock, nameof(clock));
+        }
+
+        /// <summary>
+        /// Constructs a fake scheduler with a new fake clock starting at 0 ticks.
+        /// </summary>
+        public FakeScheduler() : this(new FakeClock(0L))
+        {
+        }
+
+        // Defaulting cancellation token makes it simpler to use for some tests.
+        /// <inheritdoc />
+        public Task Delay(TimeSpan delay, CancellationToken cancellationToken = default(CancellationToken)) =>
+            AddTimer(Clock.GetCurrentDateTimeUtc() + delay, cancellationToken);
```

Comment:
I don't *think* this should be a problem - because of the `while` loop waiting for the monitor. The intention is that we don't let the clock move on until the system has been idle for a while. So it doesn't matter if `firstCurrent + 100` is queued first - so long as the `firstCurrent + 1` call is queued within `IdleTimeBeforeAdvancing`, we'll put it to the front of the queue - so we advance to `firstCurrent + 1`, not `firstCurrent + 100`.

If that description doesn't mesh with the code you're seeing, please shout!

The important point is the `while (_monitor.Wait(...))` loop though - that's the bit that should only let us proceed when the system is stable. If that's broken, we're in trouble.

Reply:
Yes, exactly, I think the problem might be in the loop, but I didn't explain myself clearly. The `Monitor.PulseAll(_monitor)` call on the `AddTimer` method will move the inner cycle thread (the only one on the wait queue) to the ready queue, but it doesn't mean the the inner cycle thread goes to the front of the ready queue, there might be some other `Delay` threads in front and there might even come some other `Delay` threads after. What this means is that even if the inner cycle thread received the Pulse before `IdleTimeBeforeAdvancing` elapsed it might well get the lock after `IdleTimeBeforeAdvancing` breaking the while loop before the system is stable, that is, before there were no Delays added during a whole `IdleTimeBeforeAdvancing`. I think.

Reply:
Even if there are lots of threads waiting on the monitor, they shouldn't be doing much. There could be other AddTimer methods waiting to add their timers, but they should be pretty quick. Given that we're allowing 100ms for everything to become steady, I'd expect that to be enough - especially as we should only have about 20 things scheduling delays. Going through the linked list *should* be quick.

Do you think it would be worth separating out the lock for the linked list from the lock just for pulsing? That would take a little bit of thought, but should be feasible.

Or I may still not be understanding you properly...

(Off to do conference things now - will have another look later on. I'll see if I can reproduce it locally too, as that would really help. I may be able to do that with shorter timeouts.)

Reply:
I think we are understanding each other.

And yes, 20 things schedulling delays should be quick, although I think we sometimes have 40 things (the 20 begin transactions from the first batch and the 20 create sessions from the second batch), but I agree that 100ms _should_ be enough even when there are 40.

> Do you think it would be worth separating out the lock for the linked list from the lock just for pulsing? That would take a little bit of thought, but should be feasible.

That could work, it would guarantee that the inner loop wakes as soon as it is pulsed, so way before `IdleTimeBeforeAdvancing` has elapsed. I'm trying something with a counter, basically counting all the delays that have come in vs. all the delays we have added to the linkedList and only breaking the while loop when that is zero (after waiting). I'll send you an alternative PR if I see that working.

And I can reproduce it (most of the time), mostly for tests that don't start acquiring sessions on the pool constructor and setting `IdleTimeBeforeAdvancing` to 25ms (our machines are way too fast, it never fails with 100ms).

The only alternative I see that can make the tests fail is that effectively no delay is schedulled in `IdleTimeBeforeAdvancing` even though the test is not done yet. That can be either

- because of the test itself doing "a lot" inside the `client.Scheduler.RunAsync` but not calling anything that actually schedulles a delay (I haven't seen a test like that).
- or because none of the threads get to the schedulling a delay point, which, by looking at the `TargettedSessionPool` code, seems improbable, even though all the (20 or 40) threads might be competing for the pool lock to get to the "fake spanner client" at least one of them should get through in 100ms I think.

Reply:
Thanks for looking into this more :) I'm going to try reducing it to 25ms as well, and see what I can find out too. We'll get there between us :)

Reply:
Take a look at PR #2635, my commit. I think that if the problem were with the PulseAll as I've been describing, that commit should have solved it, but it didn't. I can only think that we are looking at either of the two scenarios in which no delay gets schedulled in 100ms, at which point, maybe just increasing `IdleTimeBeforeAdvancing` might fix it? If we are in a deadlock, just increasing the `IdleTimeBeforeAdvancing` won't mask it, if the problem is that everything is just taking a little bit longer on the CI machines, then, that's what `IdleTimeBeforeAdvancing` is for I think. I don't have any other theories as to why the tests are failing at this point.
The two scenarios in which no delay gets schedulled in 100ms, copy pasted from above:
- because of the test itself doing "a lot" inside the `client.Scheduler.RunAsync` but not calling anything that actually schedulles a delay (I haven't seen a test like that).
- or because none of the threads get to the schedulling a delay point, which, by looking at the `TargettedSessionPool` code, seems improbable, even though all the (20 or 40) threads might be competing for the pool lock to get to the "fake spanner client" at least one of them should get through in 100ms I think.

Reply:
I'm investigating another aspect now: the "old" GAX FakeScheduler had another Monitor.Wait call that's designed to give the test code more time to run before waiting for more actions. It feels a bit wrong to me, but I'm looking into whether putting that code back helps.